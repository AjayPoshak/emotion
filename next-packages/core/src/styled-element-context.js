// @flow
import * as React from 'react'

let StyledElementsContext = React.createContext({})

let StyledElementsProvider = ({ components, children }) => (
  <StyledElementsContext.Provider value={components}>
    {children}
  </StyledElementsContext.Provider>
)

let withStyledElementsContext = function withStyledElementsContext<Props>(
  func: (props: Props, context: StyledElementsContextType) => React.Node
): React.StatelessFunctionalComponent<Props> {
  return (props: Props) => (
    <StyledElementsContext.Consumer>
      {(contextComponents: StyledElementsContextType) => {
        return func(props, contextComponents)
      }}
    </StyledElementsContext.Consumer>
  )
}

let consumeStyledElementsContext = (
  func: StyledElementsContextType => React.Node
) => {
  return <StyledElementsContext.Consumer>{func}</StyledElementsContext.Consumer>
}

export {
  StyledElementsProvider,
  consumeStyledElementsContext,
  withStyledElementsContext
}
