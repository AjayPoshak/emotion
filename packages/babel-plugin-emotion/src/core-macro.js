// @flow
import { createMacro } from 'babel-plugin-macros'

export default createMacro(({ references, state, babel }) => {
  throw new Error('something')
})
