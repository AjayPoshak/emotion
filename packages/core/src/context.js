// @flow
import { isBrowser, type EmotionCache } from '@emotion/utils'
import * as React from 'react'
import createCache from '@emotion/cache'

let EmotionCacheContext = React.createContext(isBrowser ? createCache() : null)

export let useContext: <Value>(
  context: React$Context<Value>
) => Value = (React: any).useContext

export let useState: <State>(
  initialState: (() => State) | State
) => [State, (State) => void] = (React: any).useState

export let ThemeContext = React.createContext<Object>({})
export let CacheProvider: React.ComponentType<{ value: EmotionCache }> =
  // $FlowFixMe
  EmotionCacheContext.Provider

let forwardRef: <Props>(
  render: (props: Props, ref: any) => React.Node
) => React.StatelessFunctionalComponent<Props> = (React: any).forwardRef

let withEmotionCache = function withEmotionCache<Props, Ref: React.Ref<*>>(
  func: (props: Props, cache: EmotionCache, ref: Ref) => React.Node
): React.StatelessFunctionalComponent<Props> {
  return forwardRef((props: Props, ref: Ref) => {
    let cache = useContext(EmotionCacheContext)

    return func(props, cache, ref)
  })
}

if (!isBrowser) {
  withEmotionCache = function withEmotionCache<Props>(
    func: (props: Props, cache: EmotionCache) => React.Node
  ): React.StatelessFunctionalComponent<Props> {
    return (props: Props) => {
      let cache = useContext(EmotionCacheContext)
      if (cache === null) {
        // yes, we're potentially creating this on every render
        // it doesn't actually matter though since it's only on the server
        // so there will only every be a single render
        // that could change in the future because of suspense and etc. but for now,
        // this works and i don't want to optimise for a future thing that we aren't sure about
        cache = createCache()
        return (
          <EmotionCacheContext.Provider value={cache}>
            {func(props, cache)}
          </EmotionCacheContext.Provider>
        )
      } else {
        return func(props, cache)
      }
    }
  }
}

export { withEmotionCache }
