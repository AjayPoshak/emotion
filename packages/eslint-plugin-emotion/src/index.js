exports.rules = {
  'import-from-emotion': require('./rules/import-from-emotion'),
  'no-vanilla': require('./rules/no-vanilla'),
  'syntax-preference': require('./rules/syntax-preference'),
  styled: require('eslint-plugin-emotion/src/rules/styled-import'),
  jsx: require('eslint-plugin-emotion/src/rules/jsx-import')
}
