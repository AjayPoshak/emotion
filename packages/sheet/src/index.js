// @flow
/*

Based off glamor's StyleSheet, thanks Sunil ❤️

high performance StyleSheet for css-in-js systems

- uses multiple style tags behind the scenes for millions of rules
- uses `insertRule` for appending in production for *much* faster performance

// usage

import { StyleSheet } from '@emotion/sheet'

let styleSheet = new StyleSheet({ key: '', container: document.head })

styleSheet.insert('#box { border: 1px solid red; }')
- appends a css rule into the stylesheet

styleSheet.flush()
- empties the stylesheet of all its contents

*/

// $FlowFixMe
function sheetForTag(tag: HTMLStyleElement): CSSStyleSheet {
  if (tag.sheet) {
    // $FlowFixMe
    return tag.sheet
  }

  // this weirdness brought to you by firefox
  /* istanbul ignore next */
  for (let i = 0; i < document.styleSheets.length; i++) {
    if (document.styleSheets[i].ownerNode === tag) {
      // $FlowFixMe
      return document.styleSheets[i]
    }
  }
}

export type Options = {
  nonce?: string,
  key: string,
  container: Element,
  speedy: boolean
}

type Sheet = {
  tags: Array<HTMLStyleElement>,
  ctr: number,
  before: null | Element,
  opts: {
    nonce?: string,
    key: string,
    container: Element,
    speedy: boolean
  }
}

export let createSheet = (opts: Options): Sheet => {
  let sheet = {
    tags: [],
    ctr: 0,
    before: null,
    opts
  }
  return sheet
}

export let insert = (sheet: Sheet, rule: string) => {
  // the max length is how many rules we have per style tag, it's 65000 in speedy mode
  // it's 1 in dev because we insert source maps that map a single rule to a location
  // and you can only have one source map per style tag
  let tags = sheet.tags
  let isSpeedy = sheet.opts.speedy
  if (sheet.ctr % (isSpeedy ? 65000 : 1) === 0) {
    let newTag = createStyleElement(sheet.opts)
    let before
    if (tags.length === 0) {
      before = sheet.before
    } else {
      before = tags[tags.length - 1].nextSibling
    }
    sheet.opts.container.insertBefore(newTag, before)
    tags.push(newTag)
  }
  const tag = tags[tags.length - 1]

  if (isSpeedy) {
    const sheet = sheetForTag(tag)
    try {
      // this is a really hot path
      // we check the second character first because having "i"
      // as the second character will happen less often than
      // having "@" as the first character
      let isImportRule = rule.charCodeAt(1) === 105 && rule.charCodeAt(0) === 64
      // this is the ultrafast version, works across browsers
      // the big drawback is that the css won't be editable in devtools
      sheet.insertRule(
        rule,
        // we need to insert @import rules before anything else
        // otherwise there will be an error
        // technically this means that the @import rules will
        // _usually_(not always since there could be multiple style tags)
        // be the first ones in prod and generally later in dev
        // this shouldn't really matter in the real world though
        // @import is generally only used for font faces from google fonts and etc.
        // so while this could be technically correct then it would be slower and larger
        // for a tiny bit of correctness that won't matter in the real world
        isImportRule ? 0 : sheet.cssRules.length
      )
    } catch (e) {
      if (process.env.NODE_ENV !== 'production') {
        console.warn(
          `There was a problem inserting the following rule: "${rule}"`,
          e
        )
      }
    }
  } else {
    tag.appendChild(document.createTextNode(rule))
  }
  sheet.ctr++
}

export function flush(sheet: Sheet) {
  let tag
  while ((tag = sheet.tags.shift()) !== undefined) {
    // $FlowFixMe
    tag.parentNode.removeChild(tag)
  }
  sheet.ctr = 0
}

function createStyleElement(options: {
  key: string,
  nonce: string | void
}): HTMLStyleElement {
  let tag = document.createElement('style')
  tag.setAttribute('data-emotion', options.key)
  if (options.nonce !== undefined) {
    tag.setAttribute('nonce', options.nonce)
  }
  tag.appendChild(document.createTextNode(''))
  return tag
}
