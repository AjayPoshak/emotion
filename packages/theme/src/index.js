// @flow
import * as React from 'react'
import { isBrowser } from '@emotion/utils'

let canUseCSSVars =
  isBrowser &&
  window.CSS &&
  window.CSS.supports &&
  window.CSS.supports('--test', 0)

// support nested objects and arrays
type ThemeType = {
  [key: string]: string
}

type Ret<Theme> = {
  Consumer: React.ComponentType<{
    children: Theme => React.Node
  }>,
  consume: <Props>(
    render: (props: Props, theme: Theme, ref: React.Ref<*>) => React.Node
  ) => React.ComponentType<Props>,
  Provider: React.ComponentType<{
    theme: Theme,
    children: React.Node,
    supportsCSSVariables?: boolean
  }>,
  Extender: React.ComponentType<{ children: React.Node }>
}

export let createTheme = <Theme: ThemeType>(
  defaultTheme: Theme
): Ret<Theme> => {
  let keys = Object.keys(defaultTheme)
  let cssVarTheme = keys.reduce((val, key) => {
    val[key] = `--theme-${key}`
    return val
  }, {})

  let RawThemeContext = React.createContext(defaultTheme)

  let Context = React.createContext(defaultTheme)
  // $FlowFixMe this isn't just to get flow to be quiet, i actually want to fix this because i think flow might be right
  let cssVarUsageTheme: Theme = keys.reduce((val, key) => {
    val[key] = `var(--theme-${key})`
    return val
  }, {})

  type ProviderProps = {
    theme: Theme,
    children: React.Node,
    supportsCSSVariables?: boolean
  }

  let Provider = (props: ProviderProps) => {
    if (canUseCSSVars) {
      return (
        <RawThemeContext.Provider value={props.theme}>
          <Context.Provider value={cssVarUsageTheme}>
            <div
              style={keys.reduce((val, key) => {
                val[cssVarTheme[key]] = props.theme[key]
                return val
              }, {})}
            >
              {props.children}
            </div>
          </Context.Provider>
        </RawThemeContext.Provider>
      )
    }
    return (
      <Context.Provider value={props.theme}>{props.children}</Context.Provider>
    )
  }
  if (!isBrowser) {
    let SupportsCSSVarsContext = React.createContext(false)
    Provider = props => (
      <SupportsCSSVarsContext.Consumer>
        {supportsCSSVarsFromContext => {
          let supportsCSSVars =
            props.supportsCSSVariables === undefined
              ? supportsCSSVarsFromContext
              : props.supportsCSSVariables
          return (
            <SupportsCSSVarsContext.Provider value={supportsCSSVars}>
              {supportsCSSVars ? (
                <Context.Provider value={cssVarUsageTheme}>
                  <div
                    style={keys.reduce((val, key) => {
                      val[cssVarTheme[key]] = props.theme[key]
                      return val
                    }, {})}
                  >
                    {props.children}
                  </div>
                </Context.Provider>
              ) : (
                <Context.Provider value={props.theme}>
                  {props.children}
                </Context.Provider>
              )}
            </SupportsCSSVarsContext.Provider>
          )
        }}
      </SupportsCSSVarsContext.Consumer>
    )
  }

  let Extender = (props: { children: React.Node }) => {
    if (canUseCSSVars) {
      return (
        <RawThemeContext.Consumer>
          {theme => (
            <div
              style={keys.reduce((val, key) => {
                val[cssVarTheme[key]] = theme[key]
                return val
              }, {})}
            >
              {props.children}
            </div>
          )}
        </RawThemeContext.Consumer>
      )
    }
    return props.children
  }

  let consume = <Props>(
    render: (props: Props, theme: Theme, ref: React.Ref<*>) => React.Node
  ): React.ComponentType<Props> => {
    // $FlowFixMe
    return React.forwardRef((props, ref) => {
      return (
        <Context.Consumer>
          {theme => {
            return render(props, theme, ref)
          }}
        </Context.Consumer>
      )
    })
  }

  return {
    Provider,
    Consumer: Context.Consumer,
    consume,
    Extender
  }
}
