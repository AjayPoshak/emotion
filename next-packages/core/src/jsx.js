// @flow
import * as React from 'react'
import { consume } from './context'
import { consumeStyledElementsContext } from './styled-element-context'
import { getRegisteredStyles, insertStyles, isBrowser } from '@emotion/utils'
import { serializeStyles } from '@emotion/serialize'
import { testAlwaysTrue, pickAssign } from '@emotion/styled-base/src/utils'

// $FlowFixMe
export const jsx: typeof React.createElement = function(
  type: React.ElementType,
  props: Object
) {
  // if (props == null || props.css == null) {
  //   // $FlowFixMe
  //   return React.createElement.apply(undefined, arguments)
  // }

  if (
    props &&
    props.css &&
    typeof props.css === 'string' &&
    process.env.NODE_ENV !== 'production' &&
    // check if there is a css declaration
    props.css.indexOf(':') !== -1
  ) {
    throw new Error(
      `Strings are not allowed as css prop values, please wrap it in a css template literal from '@emotion/css' like this: css\`${
        props.css
      }\``
    )
  }

  return consume(themeContext =>
    consumeStyledElementsContext(styledElementsContext => {
      const newProps = {}
      let registeredStyles = []
      let className = ''

      if (styledElementsContext[type]) {
        registeredStyles.push(
          typeof styledElementsContext[type] === 'function'
            ? styledElementsContext[type](
                pickAssign(testAlwaysTrue, {}, props, {
                  theme: themeContext.theme
                })
              )
            : styledElementsContext[type]
        )
      }

      if (props && props.className !== undefined) {
        className = getRegisteredStyles(
          themeContext.registered,
          registeredStyles,
          props.className
        )
      }

      if (props && props.css) {
        registeredStyles.push(
          typeof props.css === 'function'
            ? props.css(themeContext.theme)
            : props.css
        )
      }

      const serialized = serializeStyles(
        themeContext.registered,
        registeredStyles
      )

      const rules = insertStyles(
        themeContext,
        serialized,
        typeof type === 'string'
      )

      className += `${themeContext.key}-${serialized.name}`

      if (props) {
        for (let key in props) {
          if (
            Object.prototype.hasOwnProperty.call(props, key) &&
            key !== 'css'
          ) {
            newProps[key] = props[key]
          }
        }
      }

      newProps.className = className

      let argsLength = arguments.length
      let createElementArgArray = new Array(argsLength)

      createElementArgArray[1] = newProps
      createElementArgArray[0] = type

      for (let i = 2; i < argsLength; i++) {
        createElementArgArray[i] = arguments[i]
      }

      // $FlowFixMe
      const ele = React.createElement.apply(undefined, createElementArgArray)
      if (!isBrowser && rules !== undefined) {
        return (
          <React.Fragment>
            <style
              {...{
                [`data-emotion-${themeContext.key}`]: serialized.name,
                dangerouslySetInnerHTML: { __html: rules },
                nonce: themeContext.sheet.nonce
              }}
            />
            {ele}
          </React.Fragment>
        )
      }
      return ele
    })
  )
}
