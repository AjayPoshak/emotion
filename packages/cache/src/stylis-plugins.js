// @flow
// https://github.com/thysultan/stylis.js/tree/master/plugins/rule-sheet
// inlined to avoid umd wrapper and peerDep warnings/installing stylis
// since we use stylis after closure compiler

import type { StylisPlugin } from './types'
import { insert } from '@emotion/sheet'
import { type Sheet as SheetType } from '@emotion/utils'

const delimiter = '/*|*/'
const needle = delimiter + '}'

function toSheet(block) {
  if (block) {
    insert(Sheet.current, block + '}')
  }
}

export let Sheet: { current: SheetType } = {
  current: (null: any)
}

export let ruleSheet: StylisPlugin = (
  context,
  content,
  selectors,
  parents,
  line,
  column,
  length,
  ns,
  depth,
  at
) => {
  switch (context) {
    // property
    case 1: {
      switch (content.charCodeAt(0)) {
        case 64: {
          // @import
          insert(Sheet.current, content + ';')
          return ''
        }
        // charcode for l
        case 108: {
          // charcode for b
          // this ignores label
          if (content.charCodeAt(2) === 98) {
            return ''
          }
        }
      }
      break
    }
    // selector
    case 2: {
      if (ns === 0) return content + delimiter
      break
    }
    // at-rule
    case 3: {
      switch (ns) {
        // @font-face, @page
        case 102:
        case 112: {
          Sheet.current.insert(selectors[0] + content)
          return ''
        }
        default: {
          return content + (at === 0 ? delimiter : '')
        }
      }
    }
    case -2: {
      content.split(needle).forEach(toSheet)
    }
  }
}

export let removeLabel: StylisPlugin = (context, content) => {
  if (
    context === 1 &&
    // charcode for l
    content.charCodeAt(0) === 108 &&
    // charcode for b
    content.charCodeAt(2) === 98
    // this ignores label
  ) {
    return ''
  }
}
