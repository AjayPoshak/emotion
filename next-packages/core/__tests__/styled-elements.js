// @flow
/** @jsx jsx */
import 'test-utils/next-env'
import { jsx } from '@emotion/core'
import { StyledElementsProvider } from '@emotion/core/src/styled-element-context'
import ThemeProvider from '@emotion/provider'
import renderer from 'react-test-renderer'

test('Styled Elements - basic element', () => {
  const tree = renderer.create(
    <StyledElementsProvider
      components={{ div: { display: 'flex', color: 'hotpink' } }}
    >
      <div />
    </StyledElementsProvider>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('Styled Elements - nested elements', () => {
  const tree = renderer.create(
    <StyledElementsProvider
      components={{
        div: { display: 'flex', color: 'hotpink' },
        h1: { fontSize: 18 },
        button: { outline: 'none', border: 'none', backgroundColor: 'purple' }
      }}
    >
      <div>
        <h1>Hello</h1>
        <button>Click Me</button>
      </div>
    </StyledElementsProvider>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('Styled Elements - css', () => {
  const tree = renderer.create(
    <StyledElementsProvider
      components={{ div: { display: 'flex', color: 'hotpink' } }}
    >
      <div css={{ color: 'green' }} />
    </StyledElementsProvider>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('Styled Elements - prop function', () => {
  const tree = renderer.create(
    <StyledElementsProvider
      components={{ div: ({ fontSize }) => ({ fontSize: fontSize || 8 }) }}
    >
      <div fontSize={24} />
    </StyledElementsProvider>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('Styled Elements - theming', () => {
  const tree = renderer.create(
    <StyledElementsProvider components={{ div: { color: 'hotpink' } }}>
      <ThemeProvider theme={{ primary: 'green' }}>
        <div css={theme => ({ color: theme.primary })} />
      </ThemeProvider>
    </StyledElementsProvider>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('Styled Elements - prop function with theming', () => {
  const tree = renderer.create(
    <StyledElementsProvider
      components={{ div: ({ theme }) => ({ color: theme.color }) }}
    >
      <ThemeProvider theme={{ primary: 'green' }}>
        <div />
      </ThemeProvider>
    </StyledElementsProvider>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})
