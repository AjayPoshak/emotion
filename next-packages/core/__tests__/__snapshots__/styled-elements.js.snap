// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled Elements - basic element 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: hotpink;
}

<div
  className="emotion-0"
/>
`;

exports[`Styled Elements - css 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: hotpink;
  color: green;
}

<div
  className="emotion-0"
/>
`;

exports[`Styled Elements - nested elements 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: hotpink;
}

.emotion-0 {
  font-size: 18px;
}

.emotion-1 {
  outline: none;
  border: none;
  background-color: purple;
}

<div
  className="emotion-2"
>
  <h1
    className="emotion-0"
  >
    Hello
  </h1>
  <button
    className="emotion-1"
  >
    Click Me
  </button>
</div>
`;

exports[`Styled Elements - prop function 1`] = `
.emotion-0 {
  font-size: 24px;
}

<div
  className="emotion-0"
  fontSize={24}
/>
`;

exports[`Styled Elements - prop function with theming 1`] = `
<div
  className="emotion-0"
/>
`;

exports[`Styled Elements - theming 1`] = `
.emotion-0 {
  color: hotpink;
  color: green;
}

<div
  className="emotion-0"
/>
`;
